### PR TITLE
Backport testutils.ExecuteUntil to 3.5 branch

### PR DESCRIPTION
**Pull request by amit-rastogi**
As part of https://github.com/etcd-io/etcd/issues/15247, backporting https://github.com/etcd-io/etcd/pull/16822 to 3.5 branch will be done in 2 steps-

1. backport testutils.ExecuteUntil to 3.5 branch since tests/e2e/v3_lease_no_proxy_test.go makes use of testutils.ExecuteUntil
2. backport changes in #16822 to 3.5 branch

This PR addresses Step 1 above.

### CODE REVIEW

**Review by jmhbnz — COMMENTED** [review]
Hey @amit-rastogi -  Please ensure your commit is signed so the developer certificate of origin (DCO) check passes, i.e:

```bash
git rebase HEAD~1 --signoff
git push --force
```

/ok-to-test

Note: It looks like `tests/e2e/v3_lease_no_proxy_test.go` only has the one reference to `testutils.ExecuteUntil`.  As an alternative we could slightly tweak that line to instead call `executeWithTimeout` which is what other e2e tests in `release-3.5` currently do, refer example https://github.com/etcd-io/etcd/blob/release-3.5/tests/e2e/ctl_v3_grpc_test.go#L112.  I think my preference would be to keep the backport as minimal as possible to just tweak to use `executeWithTimeout` but I defer to maintainers on which approach is preferred.

**Comment by amit-rastogi** [comment]
> Hey @amit-rastogi - Please ensure your commit is signed so the developer certificate of origin (DCO) check passes, i.e:
> 
> ```shell
> git rebase HEAD~1 --signoff
> git push --force
> ```
> 
> /ok-to-test
> 
> Note: It looks like `tests/e2e/v3_lease_no_proxy_test.go` only has the one reference to `testutils.ExecuteUntil`. As an alternative we could slightly tweak that line to instead call `executeWithTimeout` which is what other e2e tests in `release-3.5` currently do, refer example https://github.com/etcd-io/etcd/blob/release-3.5/tests/e2e/ctl_v3_grpc_test.go#L112. I think my preference would be to keep the backport as minimal as possible to just tweak to use `executeWithTimeout` but I defer to maintainers on which approach is preferred.

Thanks @jmhbnz for your feedback. I'll analyze the usage of executeWithTimeout as you suggested. 

Also, had few observations related to the interface of EtcdProcessCluster in 3.5 vis-a-vis the main branch. The tests/e2e/v3_lease_no_proxy_test.go which needs to be backported references methods like WaitLeader and EndpointsGRPC which are not present in the 3.5 EtcdProcessCluster interface. Analyzing this as well.

Also, made minor change in EtcdProcessClusterConfig (tests/framework/e2e/cluster.go) in 3.5 to add the GoFailClientTimeout field used in tests/e2e/v3_lease_no_proxy_test.go.

Overall it appears backporting the tests/e2e/v3_lease_no_proxy_test.go may require adding more changes in the 3.5 e2e framework.

**Review by serathius — APPROVED** [review]

**Comment by ahrtr** [comment]
> The tests/e2e/v3_lease_no_proxy_test.go which needs to be backported references methods like WaitLeader and EndpointsGRPC which are not present in the 3.5 EtcdProcessCluster interface. Analyzing more on this.

Suggest to backport them as well, in this PR in a separate PR.

**Comment by amit-rastogi** [comment]
> > The tests/e2e/v3_lease_no_proxy_test.go which needs to be backported references methods like WaitLeader and EndpointsGRPC which are not present in the 3.5 EtcdProcessCluster interface. Analyzing more on this.
> 
> Suggest to backport them as well, in this PR in a separate PR.

Thanks @ahrtr. I'll create a separate PR for that.

**Review by ahrtr — APPROVED** [review]

**Pull request merged**
